fabtests/ubertest/uber.c

```diff
@@ -330,298 +329,127 @@ static int ft_recv_test_info(void)
-			if (ret) {
-				FT_PRINTERR("ft_sock_send", ret);
-				return ret;
-			}
-
-			if (ft_skip_info(hints, fabric_info))
-				continue;
-
-			ret = ft_transfer_subindex(subindex, &remote_idx);
-			if (ret)
-				return ret;
-
-			ret = ft_check_info(hints, fabric_info);
-			if (ret)
-				return ret;
+	struct fi_info *hints = NULL;
+	struct fi_info *info  = NULL;
+	int ret, result;
 
-			ft_fw_update_info(&test_info, fabric_info, subindex);
-			printf("Starting test %d-%d-%d: ", test_info.test_index,
-				subindex, remote_idx);
-			ft_show_test_info();
+	printf("Starting test %d:\n", test_info.test_index);
 
-			ret = ft_open_res();
-			if (ret) {
-				FT_PRINTERR("ft_open_res", ret);
-				return ret;
-			}
+	ret = ft_server_setup(hints, info);
+	if (ret)
+		return ret;
 
-			result = ft_init_test();
-			if (result)
-				continue;
-
-			result = ft_run_test();
-
-			ret = ft_sock_recv(sock, &sresult, sizeof sresult);
-			if (result && result != -FI_EIO) {
-				FT_PRINTERR("ft_run_test", result);
-				fprintf(stderr, "Node: %s\nService: %s \n",
-					test_info.node, test_info.service);
-				fprintf(stderr, "%s\n", fi_tostr(hints, FI_TYPE_INFO));
-				return -FI_EOTHER;
-			} else if (ret) {
-				FT_PRINTERR("ft_sock_recv", ret);
-				result = ret;
-				return -FI_EOTHER;
-			} else if (sresult) {
-				result = sresult;
-				if (sresult != -FI_EIO)
-					return -FI_EOTHER;
-			}
-		}
-		end_test = 1;
-		ret = ft_sock_send(sock, &end_test, sizeof end_test);
-		if (ret) {
-			FT_PRINTERR("ft_sock_send", ret);
-			return ret;
-		}
+	ret = ft_send_result(0, info);
+	if (ret)
+		return ret;
 
-		ret = ft_sock_recv(sock, &end_test, sizeof end_test);
-		if (ret) {
-			FT_PRINTERR("ft_sock_recv", ret);
-			return ret;
-		}
+	ret = ft_sock_send(sock, &test_info, sizeof test_info);
+	if (ret) {
+		FT_PRINTERR("ft_sock_send", ret);
+		return ret;
 	}
 
-	if (subindex == 1)
-		return -FI_ENODATA;
-
-	return result;
-}
-
-static int ft_server_child()
-{
-	struct fi_info *hints, *info;
-	int ret;
+	ret = ft_recv_result(info);
+	if (ret)
+		return ret;
 
-	hints = fi_allocinfo();
-	if (!hints)
-		return -FI_ENOMEM;
+	ret = ft_init_test();
+	if (ret)
+		return ret;
 
-	ft_fw_convert_info(hints, &test_info);
-	printf("Starting test %d:\n", test_info.test_index);
+	result = ft_run_test();
 
-	ret = fi_getinfo(FT_FIVERSION, ft_strptr(test_info.node),
-			 ft_strptr(test_info.service), FI_SOURCE,
-			 hints, &info);
-	if (ret && ret != -FI_ENODATA) {
-		FT_PRINTERR("fi_getinfo", ret);
-	} else {
-		/* Temporary fix to run one set of tests, rather than
-		 * iterating over all interfaces / addresses.
-		 * TODO: Remove iteration from ft_fw_process_list_server.
-		 */
-		if (info && info->next) {
-			fi_freeinfo(info->next);
-			info->next = NULL;
-		}
+	ret = ft_sock_send(sock, &result, sizeof result);
+	if (result) {
+		FT_PRINTERR("ft_run_test", result);
+	} else if (ret) {
+		FT_PRINTERR("ft_sock_send", ret);
+		return ret;
+	}
```

Comment:
The ft_sock_send failure is fatal to the test.  The result check is informational.  Waiting to check result until after calling ft_sock_send, rather than immediately after ft_run_test returns is odd.  I don't think we need the if (result) check.  See below.